searcher.go

```diff
@@ -0,0 +1,135 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/ServiceWeaver/weaver"
+	"github.com/ServiceWeaver/weaver/metrics"
+	"github.com/rivo/uniseg"
+)
+
+// Two counters that count cache hits and misses in the Search method.
+var (
+	cacheHits = metrics.NewCounter(
+		"search_cache_hits",
+		"Number of Search cache hits",
+	)
+	cacheMisses = metrics.NewCounter(
+		"search_cache_misses",
+		"Number of Search cache misses",
+	)
+)
+
+// Searcher is an emoji search engine component.
+type Searcher interface {
+	// Search returns the set of emojis that match the provided query.
+	Search(context.Context, string) ([]string, error)
+
+	// SearchChatGPT returns the set of emojis that ChatGPT thinks match the
+	// provided query.
+	SearchChatGPT(context.Context, string) ([]string, error)
+}
+
+// searcher is the implementation of the Searcher component.
+type searcher struct {
+	weaver.Implements[Searcher]
+	cache   weaver.Ref[Cache]
+	chatgpt weaver.Ref[ChatGPT]
+}
+
+func (s *searcher) Search(ctx context.Context, query string) ([]string, error) {
+	// Try to get the emojis from the cache, but continue if it's not found or
+	// there is an error.
+	if emojis, err := s.cache.Get().Get(ctx, query); err != nil {
+		s.Logger().Error("cache.Get", "query", query, "err", err)
+	} else if len(emojis) > 0 {
+		cacheHits.Add(1)
+		return emojis, nil
+	} else {
+		cacheMisses.Add(1)
+	}
+
+	// Perform the search. First, we lowercase the query and split it into
+	// words. For example, the query "Black cat" is tokenized to the words
+	// "black" and "cat". Then, we say an emoji matches a query if every word
+	// in the query is a substring of at least one of the emoji's labels.
+	//
+	// For example, the cat emoji has labels ["animal", "cat"]. It does NOT
+	// match the "Black cat" query because "black" is not a substring of either
+	// "animal" or "cat". The black cat emoji, on the other hand, has labels
+	// ["animal", "black", "cat"] and therefore does match the query "Black
+	// cat".
+	//
+	// We iterate over all emojis and return the ones that match the query.
+	words := strings.Fields(strings.ToLower(query))
+	var results []string
+	for emoji, labels := range emojis {
+		if matches(labels, words) {
+			results = append(results, emoji)
+		}
+	}
+	sort.Strings(results)
```

Comment:
We chatted offline. There are two main ways we can improve the search engine:

1. Improve the emoji labels. Right now, the labels are very basic. Take 🍜, for example. This emoji is not labeled "ramen" or "noodles" which makes it very hard to search for. This approach is nice because the complexity of generating good labels is hidden from the user.
2. Improve the search algorithm. We can give precedence to full matches like "cat" matching "cat" over partial matches like "cat" matching "category". This adds complexity to the app, but we can defer to extra credit sections at the end of the workshop.

We agreed to finish the workshop first and then come back to improving the algorithm.